src/main/java/com/google/sps/servlets/EnterQueue.java

```diff
@@ -61,15 +73,59 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
             Key classKey = KeyFactory.stringToKey(classCode);
             Entity classEntity = datastore.get(txn, classKey);
 
-            ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
+            // Get date in mm/dd/yyyy format
+            DateTimeFormatter FOMATTER = DateTimeFormatter.ofPattern("MM/dd/yyyy");
+            LocalDate localDate = LocalDate.now(clock);
+            String currDate = FOMATTER.format(localDate);
+
+            Filter classVisitFilter =
+                new FilterPredicate("classCode", FilterOperator.EQUAL, classCode);
+
+            Filter dateVisitFilter = new FilterPredicate("date", FilterOperator.EQUAL, currDate);
+
+            CompositeFilter visitFilter =
+                CompositeFilterOperator.and(dateVisitFilter, classVisitFilter);
+
+            int retries2 = 10;
+            while (true) {
+              Transaction txn2 = datastore.beginTransaction();
+              try {
+                Query checkAddNew = new Query("Visit").setFilter(visitFilter);
+
+                if (datastore.prepare(checkAddNew).countEntities() == 0) {
+                  Entity newVisitEntity = new Entity("Visit");
+
+                  newVisitEntity.setProperty("classCode", classCode);
+                  newVisitEntity.setProperty("date", currDate);
+                  newVisitEntity.setProperty("numVisits", 0);
+
+                  datastore.put(txn2, newVisitEntity);
+                }
+
+                txn2.commit();
+                break;
+              } catch (ConcurrentModificationException e) {
+                if (retries2 == 0) {
+                  throw e;
+                }
+                // Allow retry to occur
+                --retries2;
+              } finally {
+                if (txn2.isActive()) {
+                  txn2.rollback();
+                }
+              }
+            }
 
-            String visitCode = (String) classEntity.getProperty("visitKey");
-            Entity visitEntity = datastore.get(txn, KeyFactory.stringToKey(visitCode));
+            Query query = new Query("Visit").setFilter(visitFilter);
```

Comment:
Could you refactor this code so that the query is run only once?

Reply:
I don't think so because the entity may not exist in the first query, so we need to re-query after creating the new entity.

Reply:
What if you did this, if the visit did not exist then you then immediately set the visit count to 1 instead of zero. If it did exist then you can retrieve it `asSingleEntity` from the original query and increment the visit count.

Reply:
Currently, before I increment visits, I check to make sure the student isn't already on the queue, so setting visit count to 1 immediately won't work unless we don't mind counting repeat visits

Reply:
Okay that makes since. Another thing to consider is that if the entity didn't exist you shouldn't have to re-query because you should be able to save the entity you just created. Then if it did exist you can use the original query.